compiler/src/dotty/tools/dotc/ast/tpd.scala

```diff
@@ -1004,7 +1004,7 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
   def sameTypes(trees: List[tpd.Tree], trees1: List[tpd.Tree]): Boolean = {
     if (trees.isEmpty) trees.isEmpty
     else if (trees1.isEmpty) trees.isEmpty
-    else (trees.head.tpe eq trees1.head.tpe) && sameTypes(trees.tail, trees1.tail)
+    else (trees.head.typeOpt eq trees1.head.typeOpt) && sameTypes(trees.tail, trees1.tail)
```

Comment:
I think I have a better fix for that one: https://github.com/lampepfl/dotty/pull/4402

Reply:
@smarter I agree. Can you push #4402 on top of this and revert my corresponding commit? Thanks!

Reply:
Never mind, I can do it myself.

Reply:
I see you have already started doing this, so please go ahead. Otherwise I could have also approved yours and done the rest myself.